### PR TITLE
feat: implement pairing

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR implements Pairing for `BN254` and `BLS12-381`.

### CODE REVIEW

File: tachyon/math/finite_fields/fp12.h [hunk 1]
```diff
@@ -233,6 +236,116 @@ class Fp12 final : public QuadraticExtensionField<Fp12<Config>> {
       return this->SquareInPlace();
     }
   }
+
+  // Return α = (α₀', α₁', α₂', α₃', α₄', α₅'), such that
+  // α = (α₀ + α₁x + α₂x² + (α₃ + α₄x + α₅x²)y) * (β₀ + β₁y + β₄xy)
+  Fp12& MulInPlaceBy034(const Fp2Ty& c0, const Fp2Ty& c3, const Fp2Ty& c4) {
```
Comment (chokobole) [inline tachyon/math/finite_fields/fp12.h:242]:
```suggestion
  Fp12& MulInPlaceBy034(const Fp2Ty& beta0, const Fp2Ty& beta3, const Fp2Ty& beta4) {
```

File: tachyon/math/elliptic_curves/bn/g2_prepared.h [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    if (q.infinity()) {
+      return {};
+    } else {
+      G2Projective<Config> r(q.x(), q.y(), Fp2Ty::One());
+
+      EllCoeffs ell_coeffs;
+      // NOTE(chokobole): Assuming |Config::kAteLoopCount| consists of elements
+      // from [-1, 0, 1], representing coefficients for -1 and 1, we reserve
+      // space in |ell_coeffs| to accommodate two-thirds of the array size.
+      ell_coeffs.reserve(std::size(Config::kAteLoopCount) * 2 / 3);
```
Comment (chokobole) [inline tachyon/math/elliptic_curves/bn/g2_prepared.h:46]: Should be fixed

File: tachyon/math/finite_fields/fp12.h [hunk 3]
```diff
@@ -233,6 +236,118 @@ class Fp12 final : public QuadraticExtensionField<Fp12<Config>> {
...
+  // Return α = (α₀', α₁', α₂', α₃', α₄', α₅'), such that
+  // α = (α₀ + α₁x + α₂x² + (α₃ + α₄x + α₅x²)y) * (β₀ + β₃y + β₄xy)
+  Fp12& MulInPlaceBy034(const Fp2Ty& beta0, const Fp2Ty& beta3,
+                        const Fp2Ty& beta4) {
+    // clang-format off
+    // α = (α₀ + α₁x + α₂x² + (α₃ + α₄x + α₅x²)y) * (β₀ + β₃y + β₄xy)
+    //   = (α₀β₀ + α₃β₃p + α₅β₄pq) + (α₁β₀ + α₄β₃p + α₃β₄p)x + (α₂β₀ + α₅β₃p + α₄β₄p)x² +
+    //     (α₀β₃ + α₂β₄q + α₃β₀ + (α₀β₄ + α₁β₃ + α₄β₀)x + (α₁β₄ + α₂β₃ + α₅β₀)x²)y
+    //   = (α₃β₃ + α₅β₄q + (α₄β₃ + α₃β₄)x + (α₅β₃ + α₄β₄)x²)p + α₀β₀ + α₁β₀x + α₂β₀x²
+    //     (α₀β₃ + α₂β₄q + α₃β₀ + (α₀β₄ + α₁β₃ + α₄β₀)x + (α₁β₄ + α₂β₃ + α₅β₀)x²)y,
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/fp12.h:249]:
```suggestion
    //   = (α₀β₀ + α₃β₃p + α₅β₄pq) + (α₁β₀ + α₃β₄p + α₄β₃p)x + (α₂β₀ + α₄β₄p + α₅β₃p)x² +
    //     (α₀β₃ + α₂β₄q + α₃β₀ + (α₀β₄ + α₁β₃ + α₄β₀)x + (α₁β₄ + α₂β₃ + α₅β₀)x²)y
    //   = (α₃β₃ + α₅β₄q + (α₃β₄ + α₄β₃)x + (α₄β₄ + α₅β₃)x²)p + α₀β₀ + α₁β₀x + α₂β₀x²
    //     (α₀β₃ + α₂β₄q + α₃β₀ + (α₀β₄ + α₁β₃ + α₄β₀)x + (α₁β₄ + α₂β₃ + α₅β₀)x²)y,
```

File: tachyon/math/finite_fields/fp12.h [hunk 4]
```diff
@@ -233,6 +236,118 @@ class Fp12 final : public QuadraticExtensionField<Fp12<Config>> {
...
+    //      where p = y² and q = x³
+    // clang-format on
+
+    // a = α₀β₀ + α₁β₀x + α₂β₀x²
+    Fp6Ty a = this->c0_ * beta0;
+    // b = (α₃ + α₄x + α₅x²) * (β₃ + β₄x)
+    //   = (α₃β₃ + α₅β₄q) + (α₄β₃ + α₃β₄)x + (α₅β₃ + α₄β₄)x², where q = x³
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/fp12.h:256]:
```suggestion
    //   = (α₃β₃ + α₅β₄q) + (α₃β₄ + α₄β₃)x + (α₄β₄ + α₅β₃)x², where q = x³
```

File: tachyon/math/finite_fields/fp12.h [hunk 5]
```diff
@@ -233,6 +236,118 @@ class Fp12 final : public QuadraticExtensionField<Fp12<Config>> {
...
+  // Return α = (α₀', α₁', α₂', α₃', α₄', α₅'), such that
+  // α = (α₀ + α₁x + α₂x² + (α₃ + α₄x + α₅x²)y) * (β₀ + β₁x + β₄xy)
+  Fp12& MulInPlaceBy014(const Fp2Ty& beta0, const Fp2Ty& beta1,
+                        const Fp2Ty& beta4) {
+    // clang-format off
+    // α = (α₀ + α₁x + α₂x² + (α₃ + α₄x + α₅x²)y) * (β₀ + β₁x + β₄xy)
+    //   = (α₀β₀ + α₂β₁q + α₅β₄pq) + (α₁β₀ + α₀β₁ + α₃β₄p)x + (α₂β₀ + α₁β₁ + α₄β₄p)x² +
+    //     (α₂β₄q + α₃β₀ + α₅β₁q + (α₀β₄ + α₃β₁ + α₄β₀)x + (α₁β₄ + α₄β₁ + α₅β₀)x²)y
+    //   = (α₅β₄q + α₃β₄x + α₄β₄x²)p + (α₀β₀ + α₂β₁q) + (α₁β₀ + α₀β₁)x + (α₂β₀ + α₁β₁)x² +
+    //     (α₂β₄q + α₃β₀ + α₅β₁q + (α₀β₄ + α₃β₁ + α₄β₀)x + (α₁β₄ + α₄β₁ + α₅β₀)x²)y,
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/fp12.h:300]:
```suggestion
    //   = (α₀β₀ + α₂β₁q + α₅β₄pq) + (α₀β₁ + α₁β₀ + α₃β₄p)x + (α₁β₁ + α₂β₀ + α₄β₄p)x² +
    //     (α₂β₄q + α₃β₀ + α₅β₁q + (α₀β₄ + α₃β₁ + α₄β₀)x + (α₁β₄ + α₄β₁ + α₅β₀)x²)y
    //   = (α₅β₄q + α₃β₄x + α₄β₄x²)p + (α₀β₀ + α₂β₁q) + (α₀β₁ + α₁β₀)x + (α₁β₁ + α₂β₀)x² +
    //     (α₂β₄q + α₃β₀ + α₅β₁q + (α₀β₄ + α₃β₁ + α₄β₀)x + (α₁β₄ + α₄β₁ + α₅β₀)x²)y,
```

File: tachyon/math/finite_fields/fp12.h [hunk 6]
```diff
@@ -233,6 +236,118 @@ class Fp12 final : public QuadraticExtensionField<Fp12<Config>> {
...
+    //      where p = y² and q = x³
+    // clang-format on
+
+    // c0 = (α₅β₄q + α₃β₄x + α₄β₄x²)p +
+    //      (α₀β₀ + α₂β₁q) + (α₁β₀ + α₀β₁)x + (α₂β₀ + α₁β₁)x²,
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/fp12.h:305]:
```suggestion
    //      (α₀β₀ + α₂β₁q) + (α₀β₁ + α₁β₀)x + (α₁β₁ + α₂β₀)x²,
```

File: tachyon/math/finite_fields/fp12.h [hunk 7]
```diff
@@ -233,6 +236,118 @@ class Fp12 final : public QuadraticExtensionField<Fp12<Config>> {
...
+    //      where p = y² and q = x³
+    // c1 = (α₃β₀ + (α₂β₄ + α₅β₁)q) +
+    //      (α₀β₄ + α₃β₁ + α₄β₀)x +
+    //      (α₁β₄ + α₄β₁ + α₅β₀)x², where q = x³
+
+    // a = (α₀ + α₁x + α₂x²) * (β₀ + β₁x)
+    //   = (α₀β₀ + α₂β₁q) + (α₁β₀ + α₀β₁)x + (α₂β₀ + α₁β₁)x², where q = x³
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/fp12.h:312]:
```suggestion
    //   = (α₀β₀ + α₂β₁q) + (α₀β₁ + α₁β₀)x + (α₁β₁ + α₂β₀)x², where q = x³
```

File: tachyon/math/elliptic_curves/bn/bn_curve.h [hunk 8]
```diff
@@ -22,6 +44,80 @@ class BNCurve {
     return f;
   }
 
+  // TODO(chokobole): Leave a comment to help understand readers.
+  template <typename G1AffinePointContainer, typename G2PreparedContainer>
+  static Fp12Ty MultiMillerLoop(const G1AffinePointContainer& a,
+                                const G2PreparedContainer& b) {
+    using EllCoeffs = typename bn::G2Prepared<Config>::EllCoeffs;
+
+    class Pair {
```
Comment (Insun35) [inline tachyon/math/elliptic_curves/bn/bn_curve.h:58]: Isn't it better to define the class outside of the function for readability?

File: tachyon/math/elliptic_curves/bn/bn_curve.h [hunk 9]
```diff
@@ -22,6 +44,80 @@ class BNCurve {
...
+          for (const Pair& pair : pairs) {
+            Ell(f, pair.NextEllCoeff(), pair.g1());
+          }
+        }
+      }
+      return f;
+    };
+
+    std::vector<Fp12Ty> results =
+        base::ParallelizeMapByChunkSize(pairs, 4, callback);
```
Comment (Insun35) [inline tachyon/math/elliptic_curves/bn/bn_curve.h:107]: Is there any reason to fix the chunk size to 4?

File: tachyon/math/elliptic_curves/bls12/bls12_curve.h [hunk 10]
```diff
@@ -22,6 +46,75 @@ class BLS12Curve {
     return f;
   }
 
+  // TODO(chokobole): Leave a comment to help understand readers.
+  template <typename G1AffinePointContainer, typename G2PreparedContainer>
+  static Fp12Ty MultiMillerLoop(const G1AffinePointContainer& a,
+                                const G2PreparedContainer& b) {
+    using EllCoeffs = typename bls12::G2Prepared<Config>::EllCoeffs;
+
+    class Pair {
```
Comment (Insun35) [inline tachyon/math/elliptic_curves/bls12/bls12_curve.h:60]: Isn't it better to define the class outside of the function? 
Also, the name `Pair` seems a bit ambiguous. I would suggest `G1EllCoeffsPair` or something.

File: tachyon/math/elliptic_curves/bn/bn_curve.h (hunk 9 above)
Comment (chokobole) [reply tachyon/math/elliptic_curves/bn/bn_curve.h:107]: It's a constant! https://github.com/arkworks-rs/algebra/blob/ee0a4078b67037b4c4d22d330113a5d7c28fd6ce/ec/src/models/bn/mod.rs#L66

File: tachyon/math/elliptic_curves/bls12/bls12_curve.h (hunk 10 above)
Comment (chokobole) [reply tachyon/math/elliptic_curves/bls12/bls12_curve.h:60]: Why do you think it's better to move it outside the function ? If we move it, then I agree that `Pair` seems ambiguous. but if we maintain it inside the function, then i think it's okay!

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Review by Insun35 — APPROVED** [review]
LGTM

**Review by TomTaehoonKim — COMMENTED** [review]
LGTM

**Pull request merged**